extensions/data-plane/data-plane-http/src/main/java/org/eclipse/edc/connector/dataplane/http/pipeline/HttpRequestParams.java

```diff
@@ -81,15 +75,29 @@ private RequestBody createRequestBody(@Nullable Supplier<InputStream> bodySuppli
      *
      * @return The URL.
      */
-    private String toUrl() {
-        var url = baseUrl;
+    private HttpUrl toUrl() {
+        var parsed = HttpUrl.parse(baseUrl);
+        Objects.requireNonNull(parsed, "Failed to parse baseUrl: " + baseUrl);
+        var builder = parsed.newBuilder();
         if (path != null && !path.isBlank()) {
-            url += "/" + path;
+            builder.addPathSegments(path);
+        }
+        if (queryParams != null) {
+            parseQueryParams(queryParams).forEach(builder::addQueryParameter);
         }
-        if (queryParams != null && !queryParams.isBlank()) {
-            url += "?" + queryParams;
+
+        return builder.build();
+    }
+
+    private static Map<String, String> parseQueryParams(String s) {
```

Comment:
instead of first encoding all query params in a single string outside of the HttpRequestParams, and then parsing them here, couldn't we modify the `HttpRequestParams.Builder` such that it takes the query params as a `Map<String, String>` and add a `queryParam(String key, String value)` method?

Reply:
done